packages/calling-stateful-client/src/CallAgentDeclarativeCommon.ts

```diff
@@ -0,0 +1,287 @@
+  convertSdkIncomingCallToDeclarativeIncomingCall,
+  isACSCall,
+  isACSCallAgent
+} from './Converter';
+import { DeclarativeIncomingCall, incomingCallDeclaratify } from './IncomingCallDeclarative';
+import { IncomingCallSubscriber } from './IncomingCallSubscriber';
+import { InternalCallContext } from './InternalCallContext';
+import { disposeAllViewsFromCall } from './StreamUtils';
+/* @conditional-compile-remove(teams-call) */
+import { DeclarativeTeamsCall, teamsCallDeclaratify } from './TeamsCallDeclarative';
+
+/**
+ * TODO: This should likely be exported?
+ *
+ * @private
+ */
+export type DeclarativeCallCommon = CallCommon & {
+  /**
+   * Stop any declarative specific subscriptions and remove declarative subscribers.
+   */
+  unsubscribe(): void;
+};
+
+/**
+ * ProxyCallAgent proxies CallAgent and saves any returned state in the given context. It will subscribe to all state
+ * updates in the CallAgent and in the contained Calls and RemoteParticipants. When dispose is called it will
+ * unsubscribe from all state updates.
+ */
+export abstract class ProxyCallAgentCommon {
+  private _context: CallContext;
+  private _internalContext: InternalCallContext;
+  private _callSubscribers: Map<CallCommon, CallSubscriber>;
+  private _incomingCallSubscribers: Map<string, IncomingCallSubscriber>;
+  private _declarativeIncomingCalls: Map<string, DeclarativeIncomingCall>;
+  private _declarativeCalls: Map<CallCommon, DeclarativeCallCommon>;
+  private _externalCallsUpdatedListeners: Set<CollectionUpdatedEvent<CallCommon>>;
+
+  constructor(context: CallContext, internalContext: InternalCallContext) {
+    this._context = context;
+    this._internalContext = internalContext;
+    this._callSubscribers = new Map<Call, CallSubscriber>();
+    this._incomingCallSubscribers = new Map<string, IncomingCallSubscriber>();
+    this._declarativeIncomingCalls = new Map<string, DeclarativeIncomingCall>();
+    this._declarativeCalls = new Map<Call, DeclarativeCall>();
+    this._externalCallsUpdatedListeners = new Set<CollectionUpdatedEvent<CallCommon>>();
+  }
+
+  protected abstract subscribe(): void;
+
+  // Unsubscribe is called when CallAgent is disposed. This should mean no more updating of existing call but we don't
+  // remove any existing state.
+  protected unregisterSubscriber = (): void => {
+    for (const [_, callSubscriber] of this._callSubscribers.entries()) {
+      callSubscriber.unsubscribe();
+    }
+    this._callSubscribers.clear();
+
+    for (const [_, incomingCallSubscriber] of this._incomingCallSubscribers.entries()) {
+      incomingCallSubscriber.unsubscribe();
+    }
+    this._incomingCallSubscribers.clear();
+    this._declarativeIncomingCalls.clear();
+
+    for (const [_, declarativeCall] of this._declarativeCalls.entries()) {
+      declarativeCall.unsubscribe();
+    }
+    this._declarativeCalls.clear();
+  };
+
+  protected abstract unsubscribe(): void;
+
+  protected callsUpdated = (event: {
```

Comment:
I don't like this.

Reply:
Me either! Updating these to the new pattern we've discussed!